crates/libs/core/src/runtime/stateful.rs

```diff
@@ -111,43 +128,132 @@ pub trait LocalReplicator: Send + Sync + 'static {
     fn get_current_progress(&self) -> crate::Result<i64>;
+
+    /// Get the first LSN, beginning of log.
+    /// Remarks:
+    /// SF uses this to determine if other replicas can catch up from this replica.
+    /// Other replica's end of log must be higher than this replica's beginning of log
+    /// in order for the other replica to catchup, otherwise SF needs to drop the other
+    /// replica (if the current replica is chosen to be primary).
     fn get_catch_up_capability(&self) -> crate::Result<i64>;
     fn abort(&self);
 }
 
-/// TODO: primary replicator has no public documentation
+// Remarks:
+// Adding a secondary into the partition involves the following steps:
+// * SF brings up an idle secondary replica S which can be empty or contain
+// (partial) previous data.
+// * build_replica is called on primary to copy data into the S.
+// * S is changed to active secondary
+// * update_catch_up_replica_set_configuration to include S in the current configuration,
+// and wait_for_catch_up_quorum are called on primary for final synchronization before SF
+// grants ReadStatus to S.
+// * SF grants ReadStatus to S, and replica build completes.
+//
+// For primary failover, all active or idle secondaries gets update_epoch() call, and new
+// primary gets the new epoch from change_role() call. Secondary should fence/reject
+// operations from the old primary with an older epoch.
+
+/// TODO: primary replicator has no public documentation, this is gathered unofficially and
+/// is subject to change/correction.
 /// IFabricPrimaryReplicator com interface wrapper.
 #[trait_variant::make(PrimaryReplicator: Send)]
 pub trait LocalPrimaryReplicator: Replicator {
     // SF calls this to indicate that possible data loss has occurred (write quorum loss),
     // returns is isStateChanged. If true, SF will re-create other secondaries.
     // The default SF impl might be a pass through to the state provider.
     async fn on_data_loss(&self, cancellation_token: CancellationToken) -> crate::Result<u8>;
+
+    // Remarks on replicator configuration:
+    // At any time the replicator can have one or two configurations. There is always a current
+    // configuration which represents the set of replicas that are participating in replication
+    // along with the current write quorum. In addition there can be a previous configuration
+    // which represents the set of replicas that were in the previous configuration.
+    // When there is both a current and previous configuration the replicator must ensure that
+    // writes are acknowledeged by a write quroum of both configurations.
+
+    /// Informs the replicator there there is a current configuration and a previous configuration.
+    /// Called on primary to inform the set of active secondary replicas that may
+    /// begin to catchup. Idle secondary replicas are not included here.
+    ///
+    /// The total number of replica marked with must_catchup will not exceed the write quorum.
```

Comment:
What is the significance of the `must_catchup` flag when provided as `true`?

Reply:
If the catchup mode is write, it is upto the replicator to pick a quorum of the replicas that includes the must_catch up replica. Wait_for_catchup_quorum complete can return if the quorum of replica has caught up, and can leave other slow replica still running in background.